lib/portlayer/exec/container.go

```diff
@@ -146,11 +149,65 @@ func (s State) String() string {
 		return "Stopping"
 	case StateStopped:
 		return "Stopped"
-
+	case StateUnknown:
+		return "Unknown"
 	}
 	return ""
 }
 
+// CurrentState returns current state.
+func (c *Container) CurrentState() State {
+	c.m.Lock()
+	defer c.m.Unlock()
+	return c.state
+}
+
+// SetState changes container state.
+func (c *Container) SetState(s State) State {
```

Comment:
Please make this test&set  as per initial comments - otherwise we're still racing on restoring existingState on error paths.